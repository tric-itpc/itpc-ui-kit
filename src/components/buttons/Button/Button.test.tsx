import React from "react"

import "@testing-library/jest-dom"
import { fireEvent, render, screen } from "@testing-library/react"

import { Button, Props } from "./index"

describe("Компонент Button", () => {
  const defaultProps: Props = {
    children: "Нажми меня",
    className: "test-class",
    disabled: false,
    onPress: jest.fn(),
    type: "button",
    variant: "primary",
  }

  test("Рендерится с дефолтными пропсами (снимок)", () => {
    const { container } = render(<Button {...defaultProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test.each(["primary", "red", "white"] as const)(
    "Рендерится с вариантом '%s' (снимок)",
    (variant) => {
      const { container } = render(
        <Button {...defaultProps} variant={variant} />
      )
      expect(container.firstChild).toMatchSnapshot()
    }
  )

  test.each(["button", "submit", "reset"] as const)(
    "Рендерится с типом кнопки '%s' (снимок)",
    (type) => {
      const { container } = render(<Button {...defaultProps} type={type} />)
      expect(container.firstChild).toMatchSnapshot()
    }
  )

  test("Корректно отображает текст внутри", () => {
    render(<Button {...defaultProps}>Тестовый контент</Button>)
    expect(screen.getByRole("button")).toHaveTextContent("Тестовый контент")
  })

  test("Применяет атрибут disabled, если он передан", () => {
    render(<Button {...defaultProps} disabled />)
    expect(screen.getByRole("button")).toBeDisabled()
  })

  test("Вызывает onPress при клике", () => {
    const onPressMock = jest.fn()
    render(<Button {...defaultProps} onPress={onPressMock} />)
    const button = screen.getByRole("button")
    fireEvent.click(button)

    expect(onPressMock).toHaveBeenCalledTimes(1)
    expect(onPressMock).toHaveBeenCalledWith(expect.any(Object))
  })

  test("Не вызывает onPress, если кнопка отключена", () => {
    const onPressMock = jest.fn()
    render(<Button {...defaultProps} onPress={onPressMock} disabled />)
    const button = screen.getByRole("button")
    fireEvent.click(button)
    expect(onPressMock).not.toHaveBeenCalled()
  })

  test.each(["primary", "red", "white"] as const)(
    "Добавляет корректный класс для варианта '%s'",
    (variant) => {
      render(<Button {...defaultProps} variant={variant} />)
      expect(screen.getByRole("button")).toHaveClass(
        `itpc-button__color_${variant}`
      )
    }
  )

  test.each(["button", "submit", "reset"] as const)(
    "Устанавливает корректный тип кнопки '%s'",
    (type) => {
      render(<Button {...defaultProps} type={type} />)
      expect(screen.getByRole("button")).toHaveAttribute("type", type)
    }
  )

  test("Передаёт дополнительные атрибуты", () => {
    render(<Button {...defaultProps} data-testid="custom-button" />)
    expect(screen.getByTestId("custom-button")).toBeInTheDocument()
    expect(screen.getByTestId("custom-button")).toHaveAttribute(
      "data-testid",
      "custom-button"
    )
  })

  test("Добавляет кастомный класс, если он передан", () => {
    render(<Button {...defaultProps} className="custom-class" />)
    const button = expect(screen.getByRole("button")).toHaveClass(
      "custom-class"
    )
  })
})
