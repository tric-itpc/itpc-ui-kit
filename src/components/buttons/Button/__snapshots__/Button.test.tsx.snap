// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент Button Рендерится с вариантом 'primary' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_primary test-class"
  type="button"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с вариантом 'red' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_red test-class"
  type="button"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с вариантом 'white' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_white test-class"
  type="button"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с дефолтными пропсами (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_primary test-class"
  type="button"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с типом кнопки 'button' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_primary test-class"
  type="button"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с типом кнопки 'reset' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_primary test-class"
  type="reset"
>
  Нажми меня
</button>
`;

exports[`Компонент Button Рендерится с типом кнопки 'submit' (снимок) 1`] = `
<button
  class="itpc-button itpc-button__color_primary test-class"
  type="submit"
>
  Нажми меня
</button>
`;
